tsl/src/chunk.c

```diff
@@ -259,3 +260,92 @@ chunk_invoke_drop_chunks(Oid relid, Datum older_than, Datum older_than_type)
 
 	return num_results;
 }
+
+/*
+ * drop_chunk_replica:
+ *
+ * This function drops a chunk on a specified data node. It then
+ * removes the metadata about the association of the chunk to this
+ * datanode on the access node.
+ */
+Datum
+drop_chunk_replica(PG_FUNCTION_ARGS)
+{
+	const char *node_name = PG_ARGISNULL(0) ? NULL : NameStr(*PG_GETARG_NAME(0));
+	Oid chunk_relid = PG_ARGISNULL(1) ? InvalidOid : PG_GETARG_OID(1);
+	List *data_nodes, *replicas;
+	ForeignServer *server;
+	Chunk *chunk;
+	ChunkDataNode *cdn;
+	char *drop_cmd;
+
+	if (!OidIsValid(chunk_relid))
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("Invalid chunk specification %u", chunk_relid)));
+
+	chunk = ts_chunk_get_by_relid(chunk_relid, false);
+
+	if (NULL == chunk)
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE), errmsg("Invalid chunk specified")));
+
+	/* It has to be a foreign table chunk */
+	if (chunk->relkind != RELKIND_FOREIGN_TABLE)
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("\"%s\" is not a valid remote chunk", get_rel_name(chunk_relid))));
+
+	server = data_node_get_foreign_server(node_name, ACL_USAGE, true, false);
+	Assert(NULL != server);
+
+	/* Early abort on missing permissions */
+	ts_hypertable_permissions_check(chunk_relid, GetUserId());
+
+	/* check that the chunk is indeed present on the specified data node */
+	cdn = ts_chunk_data_node_scan_by_chunk_id_and_node_name(chunk->fd.id,
+															server->servername,
+															CurrentMemoryContext);
+
+	if (!cdn)
+		ereport(ERROR,
+				(errcode(ERRCODE_UNDEFINED_OBJECT),
+				 errmsg("Chunk \"%s\" does not exist on data node \"%s\"",
+						get_rel_name(chunk_relid),
+						node_name)));
+
+	/* Ensure that this chunk remains replicated so as to avoid data loss */
+	replicas = ts_chunk_data_node_scan_by_chunk_id(chunk->fd.id, CurrentMemoryContext);
+
+	/*
+	 * There should be atleast one surviving replica after the deletion here.
+	 *
+	 * We could fetch the corresponding hypertable and check its
+	 * replication_factor. But the user of this function is using it
+	 * to move chunk from one datanode to another and is well aware of
+	 * the replication_factor requirements
+	 */
+	if (list_length(replicas) <= 1)
+		ereport(ERROR,
+				(errcode(ERRCODE_TS_INSUFFICIENT_NUM_DATA_NODES),
+				 errmsg("dropping last chunk could lead to data loss")));
+
+	/*
+	 * Drop chunk on the data node using a regular "DROP TABLE". CASCADE is
```

Comment:
I am not sure if the comment about `CASCADE` is fully correct. If a user created any database objects on the chunk directly, then `CASCADE` is necessary to be able to successfully execute the drop table. The drop chunk, which is called by `DROP TABLE`, only removes objects created by TimescaleDB.

The current approach is that the user should separately drop user-defined objects on the chunk before, e.g., calling this API function.

I suggest to improve the comment that it is not needed for removing database objects such as compressed chunk. Otherwise it might mislead the code reader.

Reply:
As per the FORGE environment, DNs are not supposed to be directly usable by users. If users want to still go ahead and do things on the DN then it's on them to also clean up appropriately.

Reply:
I think it is fine to elide CASCADE, especially since it is clear from the comment that we do this consciously.  We can always revisit if it turns out to be a problem.

Reply:
My note is about the code comment. Now it sounds that `CASCADE` will take care of non-TimescaleDB objects too. The comment is important for code maintainability and should help a developer to think in the correct direction and not mislead.

Reply:
`CASCADE` will indeed drop other objects associated with this chunk. That's what `CASCADE` is supposed to do.

Reply:
My suggestion is to improve the comment, so it will not mislead.

Reply:
I don't think it's our job to add a comment about users adding additional objects.. however, since you are insisting still, I am adding it...

Reply:
@nikkhils My comment is that current formulation is misleading. You might remove completely this sentence. You can just say that `DROP TABLE` will drop the chunk. What the drop chunk does can be checked in the corresponding implementation and is not necessary to describe here, imho.